### DISCUSSION TITLE
[general] is it possible to run pipeline with external parameters?

### DISCUSSION THREAD

**Discussion by michkam89** [ropensci/targets#1025] [General]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi! I am wondering, if I can run a pipeline with external parameters not defined in `_targets.R`?

I would like to do call `tar_make()` and provide parameter that will be available during pipeline run. Eg.

call:
```r
tar_make(external_params = list(param="value"))
``` 

_targets.R:
```r
library(targets)
list(x, foo(y=param))
```

Basically this would be interesting usecase when you do not now the value upfront, and you want to run the pipeline programmatically.

Thanks for any hints!

**Comment by wlandau** [comment]
Hi @michkam89,

Although I do not have plans to implement this kind of parameterization at the level of `tar_make()`, you can write your `_targets.R` file with a parameterized style.

```r
# _targets.R file
params <- list(width = 7, height = 7)

library(targets)
tar_source()
list(
  tar_target(plot, make_my_ggplot()),
  tar_target(
    file, {
      ggsave("file.png", plot, width = params$width, height = params$height)
      "file.png"
    },
    format = "file"
  )
)
```

**Reply by michkam89** [reply]
Thanks for the answer @wlandau. I'm aware of this solution, however I was thinking about providing parameters, that I do not know at the time of writing `_targets.R`.

Namely, I am running `targets` pipeline as a Kubeflow Pipelines component. Transfer of files between components there is based on filesystem and the paths are defined during the run. Eg. if I want to pass an output of one component to the following one I need to save it and then load it in another component.

To achieve this I did some walkaround that includes usage of `run.R` script.

To trigger pipeline I simply run ` Rscript run.R arg1 arg2`, where one of arguments can be input defined by user, and the other can be definition of an output defined by kubeflow.

Here are the files:

```r
# run.R file
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)
saveRDS(args, file = "kf_args")

targets::tar_make()
``` 

```r
# _targets.R
library(targets)
args <- readRDS("kf_args")

list(
  tar_target(x, foo(args[1])),
  tar_target(y, bar(x, args[2]))
)
```

Maybe it's not that elegant but it works, maybe someone will have a use of this :)
